Makefile: use ln instead of cp to save space

oharboe opened the pull request:
this starts to matter for .odb files in the gigabytes size, which will happen e.g. in exploration of mm^2 sized ASAP7 designs with lots of empty space.

Empty space on a die can be common when exploring floorplans and starting with an overgenerous placement of macros.

oharboe committed:
Makefile: use ln -fs instead of cp to save space

this starts to matter for .odb files in the gigabytes size,
which will happen e.g. in exploration of mm^2 sized
ASAP7 designs with lots of empty space.

Empty space on a die can be common when exploring floorplans
and starting with an overgenerous placement of macros.

make works with softlinks: when the date is identical to
the target, i.e. it is up to date.

Files are only copied "forward" in the flow, so
this is safe.

tar works with hardlinks and softlinks: if the file does not exist in the
archive, then the full file is archived. Subsequent references
to the same file gets a hardlink. This should see "make issue"
sizes reduced.

A symlink is overwritten when a file is opened in write & truncate
mode, thus not changing the file that the symlink points to. This
is unlike a hard link.

Relative paths are used, such that the results folder
can be moved.

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

maliberty commented on flow/Makefile:
```diff
@@ -396,11 +396,11 @@ $(RESULTS_DIR)/1_1_yosys.v: $(DONT_USE_LIBS) $(WRAPPED_LIBS) $(DONT_USE_SC_LIB)
 $(RESULTS_DIR)/1_synth.v: $(RESULTS_DIR)/1_1_yosys.v
 	mkdir -p $(RESULTS_DIR) $(LOG_DIR) $(REPORTS_DIR)
-	cp $< $@
+	ln $< $@
```
What happened to the -s?

oharboe replied:
Full explanation here: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1000#issuecomment-1515814679

maliberty commented on flow/Makefile:
```diff
@@ -736,7 +736,7 @@ $(GDS_MERGED_FILE): $(RESULTS_DIR)/6_final.def $(OBJECTS_DIR)/klayout.lyt $(GDSO
 
 $(GDS_FINAL_FILE): $(GDS_MERGED_FILE)
-	cp $^ $@
+	ln $^ $@
```
Still hard linking here.

oharboe replied:
Fixed

maliberty reviewed (changes requested):
Please look at the test failures

oharboe committed:
makefile: fix potential bug

Copy first dependency, not all of them.

Allthough there is a distinction in Makefile between $^
and $<, there's no difference in the case that is fixed
as of writing.

This commit is just to clearly show a drive-by fix, before
the next commit.

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

maliberty commented on tools/OpenROAD:
Why is this part of this PR?

oharboe commented on flow/Makefile:
```diff
@@ -743,7 +743,7 @@ $(GDS_MERGED_FILE): $(RESULTS_DIR)/6_final.def $(OBJECTS_DIR)/klayout.lyt $(GDSO
 
 $(GDS_FINAL_FILE): $(GDS_MERGED_FILE)
-	cp $< $@
+	ln -sf $(shell realpath --relative-to=$(dir $@) $<) $@
```
@maliberty The relative path is relative to the destination folder of the symlink location.
